services/job_runner.go

```diff
@@ -87,12 +87,12 @@ func (rm *jobRunner) resumeSleepingRuns() error {
 
 func (rm *jobRunner) demultiplexRuns(starterWg *sync.WaitGroup) {
 	starterWg.Done()
-	rm.demultiplexStopperWg.Add(1)
 	defer rm.demultiplexStopperWg.Done()
 	for {
 		select {
 		case <-rm.done:
 			logger.Debug("JobRunner demultiplexing of job runs finished")
+			rm.workersWg.Wait()
 			return
```

Comment:
Here is a subtle but meaningful change. One of the data races caught was `rm.workersWg` being invoked in `Stop` on the main goroutine, and `rm.workersWg.Add(1)` happening in the demultiplexRuns goroutine. By placing the `rm.workersWg.Wait()` here, we are waiting in the same goroutine as we are adding, removing the data race condition.

We still properly clean up in Stop because the subsequence `rm.demultiplexStopperWg.Wait()` is still being invoked in `Stop`.